### PR TITLE
Firebase / Show public and private channels in channel list

### PR DESCRIPTION
**Pull request by JozefCeluch**
Public and private channels are now displayed in the channel list. The channels are indexed separately in firebase and database rules were updated to ensure the correct access rights.

Nothing changes in the UI, the user does not know if the channel is private or public yet.

Paired with
@Dorvaryn and @dominicfreeston on the DB rules

### CODE REVIEW

File: Firebase/android/app/src/main/java/com/novoda/bonfire/Dependencies.java [hunk 1]
```diff
@@ -25,15 +27,19 @@ public void init(Context context) {
...
-            channelService = new FirebaseChannelService(firebaseApp);
+            loginService = new FirebaseLoginService(firebaseDatabase, firebaseAuth);
+            chatService = new FirebaseChatService(firebaseDatabase);
+            channelService = new FirebaseChannelService(firebaseDatabase);
         }
     }
 
     private boolean needsInitialisation() {
-        return loginService == null && chatService == null;
+        return loginService == null && chatService == null && channelService == null && firebaseAnalytics == null;
```
Comment (takecare) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/Dependencies.java]: wouldn't it be enough that one of these is `null`?

File: Firebase/android/app/src/main/java/com/novoda/bonfire/channel/service/FirebaseChannelService.java [hunk 2]
```diff
@@ -1,42 +1,84 @@
...
     @Override
-    public Observable<Channels> getChannels() {
-        return Observable.create(new Observable.OnSubscribe<Channels>() {
+    public Observable<Channels> getChannelsFor(User user) {
+        return Observable.zip(
+                privateChannelsFor(user),
+                publicChannels(),
+                new Func2<List<Channel>, List<Channel>, Channels>() {
+                    @Override
+                    public Channels call(List<Channel> channels, List<Channel> channels2) {
```
Comment (takecare) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/channel/service/FirebaseChannelService.java]: `channels2` 😞

File: Firebase/android/app/src/main/java/com/novoda/bonfire/login/service/FirebaseLoginService.java [hunk 3]
```diff
@@ -75,6 +78,8 @@ public void loginWithGoogle(String idToken) {
                     public void onComplete(@NonNull Task<AuthResult> task) {
                         if (task.isSuccessful()) {
                             FirebaseUser firebaseUser = task.getResult().getUser();
+                            User user = new User(firebaseUser.getUid(), firebaseUser.getDisplayName(), firebaseUser.getPhotoUrl().toString()); //Refactor this double creation
```
Comment (takecare) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/login/service/FirebaseLoginService.java]: missing `TODO`?

File: Firebase/android/core/src/main/java/com/novoda/bonfire/channel/presenter/ChannelsPresenter.java [hunk 4]
```diff
@@ -4,33 +4,60 @@
...
-        subscriptions.add(channelService.getChannels().subscribe(new Action1<Channels>() {
+        subscriptions.add(loginService.getAuthentication()
+                                  .filter(successfullyAuthenticated())
+                                  .flatMap(channelsForUser())
+                                  .subscribe(new Action1<Channels>() {
+                                      @Override
+                                      public void call(Channels channels) {
+                                          channelsDisplayer.display(channels);
+                                      }
+                                  }));
```
Comment (takecare) [inline Firebase/android/core/src/main/java/com/novoda/bonfire/channel/presenter/ChannelsPresenter.java]: new line 💃

File: Firebase/android/app/src/main/java/com/novoda/bonfire/Dependencies.java (hunk 1 above)
Comment (JozefCeluch) [inline Firebase/android/app/src/main/java/com/novoda/bonfire/Dependencies.java]: yep

**Comment by takecare** [comment]
👍

**Pull request merged**